component/newstorage/couchdb/store.go

```diff
@@ -0,0 +1,604 @@
+		}
+
+		return nil, err
+	}
+
+	tags, err := getTagsFromRawDoc(rawDoc)
+	if err != nil {
+		return nil, fmt.Errorf(failGetTagsFromRawDoc, err)
+	}
+
+	return tags, nil
+}
+
+// GetBulk fetches the values associated with the given keys.
+// If a key doesn't exist, then a nil []byte is returned for that value. It is not considered an error.
+func (s *Store) GetBulk(keys ...string) ([][]byte, error) {
+	return nil, errors.New("not implemented")
+}
+
+// Query returns all data that satisfies the expression. Expression format: TagName:TagValue.
+// If TagValue is not provided, then all data associated with the TagName will be returned.
+// For now, expression can only be a single tag Name + Value pair.
+// If no options are provided, then defaults will be used.
+func (s *Store) Query(expression string, options ...newstorage.QueryOption) (newstorage.Iterator, error) {
+	return &couchDBResultsIterator{}, errors.New("not implemented")
+}
+
+// Delete deletes the key + value pair (and all tags) associated with key.
+func (s *Store) Delete(k string) error {
+	if k == "" {
+		return errors.New("key is mandatory")
+	}
+
+	revID, err := s.getRevID(k)
+	if err != nil {
+		return fmt.Errorf(failGetRevisionID, err)
+	}
+
+	// If no revision ID is returned, then this value doesn't exist.
+	// This is not considered an error.
+	if revID == "" {
+		return nil
+	}
+
+	_, err = s.db.Delete(context.TODO(), k, revID)
+	if err != nil {
+		return fmt.Errorf("failed to delete document via client: %w", err)
+	}
+
+	return nil
+}
+
+// Batch performs multiple Put and/or Delete operations in order.
+func (s *Store) Batch(operations []newstorage.Operation) error {
+	return errors.New("not implemented")
+}
+
+// Close closes this store.
+func (s *Store) Close() error {
+	err := s.db.Close(context.Background())
+	if err != nil {
+		return fmt.Errorf("failed to close database client: %w", err)
+	}
+
+	return nil
+}
+
+func (s *Store) put(k string, value []byte) error {
+	err := backoff.Retry(func() error {
+		revID, err := s.getRevID(k)
+		if err != nil {
+			// This is an unexpected error. Return a backoff.Permanent wrapped error to prevent further retries.
+			return backoff.Permanent(fmt.Errorf(failGetRevisionID, err))
+		}
+
+		if revID != "" {
+			value = []byte(`{"_rev":"` + revID + `",` + string(value[1:]))
+		}
+
+		_, err = s.db.Put(context.Background(), k, value)
+		if err != nil {
+			if err.Error() == documentUpdateConflictErrMsgFromKivik {
+				// This means that the document was updated since we got the revision ID.
+				// Need to get the new revision ID and try again.
+				return fmt.Errorf(failPutValueViaClient, err)
+			}
+
+			err = fmt.Errorf(failPutValueViaClient, err) // Wrap error here and return below.
+		}
+
+		// If there is no error or it's something other than an update conflict, then no more retries are needed.
+		// We return a backoff.Permanent wrapped error to prevent further retries.
+		return backoff.Permanent(err)
```

Comment:
err is `nil` here, set `nil` to make it explicit

Reply:
Done